### PR TITLE
ADR: CloudEvents Error Extension

### PR DESCRIPTION
**Pull request by yordis**
This spec takes into consideration the following

- Dealing with internationalization
- Deal with unique errors in distributed system environments
- Allowing customer support systems to have more context into errors that end-users may be facing
- Documenting errors
- Structured logging
- Form/Data validation errors aggregation

## Concerns

- Should we add a key for internal metadata? Some information may not be useful to expose to some clients because they are completely related to internal contexts. How do we deal with that?

### CODE REVIEW

**Comment by yordis** [comment]
Should we agree on some metadata keys being reserved for a particular thing?

For example, use `field_name` to represent the name of the form's field that the error may be associated with.

**Comment by yordis** [comment]
What about changing the serialization of the metadata to be a two-dimensional array?

```jsonc
"metadata": [
    { "key": "field_name", "value": "username" }
    // vs.
    ["field_name", "username"]
  ]
```

**Comment by germainc** [comment]
> What about changing the serialization of the metadata to be a two-dimensional array?
> 
> ```js
> "metadata": [
>     { "key": "field_name", "value": "username" }
>     // vs.
>     ["field_name", "username"]
>   ]
> ```
> 
> We lose the information about what is what relying on programmers understand that position `0` is the key, and position `1` is the value.
> 
> Added value?

I'm not seeing any real benefit to using arrays. Key/value pairs will be a lot faster to search and retrieve values.

**Comment by yordis** [comment]
> I'm not seeing any real benefit to using arrays. Key/value pairs will be a lot faster to search and retrieve values.

If we use a Map, we lost the ability to have multiple metadata under the same key, which takes a look at the latest version because I moved the information about the subject (which could be a form field) outside.

Also, I completely redo this to be on top of CloudEvents.

**Comment by yordis** [comment]
I talked to @clemensv about proposing this ADR to CloudEvents itself, so I cleaned it up to have such a conversation.

I am gonna leave, `field_name` "out of this spec", and write another ADR just for that, personally.

I think that is super important, and I would propose using `subject` with a format of [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03) to allow people to mark particular error to a form/data validation. (I would like to hear the feedback from CloudEvents folks)

**Comment by yordis** [comment]
In hold, waiting for https://github.com/cloudevents/spec/pull/741 review

File: adrs/6860374633/README.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+This ADR is a specification for describing error data in standard formats to
+provide interoperability across services, platforms, and systems.
+
+## Context Attributes Extension
+
+### dataschema
+
+- Constraints:
+  - REQUIRED
+  - MUST be <https://github.com/straw-hat-team/adr/tree/master/adrs/6860374633/v1/dataschema.json>
```
Comment (manuphatak) [inline adrs/6860374633/README.md:43]: This is a broken link

Comment (yordis) [reply adrs/6860374633/README.md:43]: yeah it is until we merge the ADR, I am not sure how to solve the issue since we need an URL to it.

**Comment by yordis** [comment]
@germainc any concerns about the spec? 

Although this is related to cloud event, trying to be able to generate cloud event errors will force you to add metadata that covers the details that we discussed it.

Now we Go new features around the errors, we could still keep some HTTP concerns, but just not the information metadata, or at least I understood kind of that watching some video, we can sync up in person.

**Pull request merged**